### PR TITLE
TandemFoil Paper: 3L/224d intermediate width

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

3L/224d is an intermediate point between the current 192d champion (PR #3025, field_mse=0.002383) and the wider 256d being tested by robin (PR #3124). The 224d hidden dimension gives ~37% more parameters than 192d (224²×3 vs 192²×3) while staying below the 256d width.

Key motivation for 224d specifically:
- 4 attention heads divide cleanly into 224d (56 dim per head), giving a cleaner per-head representation than 192d/3=64 (also clean, but 224/4=56 may capture different feature partitioning)
- May capture more pressure field detail than 192d without the instability risk of wider models
- 4L/192d caused pressure overflow (Infinity) — so layer depth is risky; width is the safer axis to explore
- This is a direct ablation point in the width sweep (192 → 224 → 256)

If 224d beats 192d, we have evidence that width is a reliable lever for TFP. If it diverges or underperforms, it sharpens the story that 192d is at or near the sweet spot.

## Instructions

Train a 3-layer model with hidden dim 224 and 4 attention heads on the TandemFoil Paper dataset. All other hyperparameters mirror the champion configuration from PR #3025.

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 \
  --model-hidden-dim 224 --model-heads 4 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb-group tfp-width-sweep
```

**What to report:**
- `val_primary/field_mse` — primary metric, beat 0.002383 to win
- Whether pressure stays finite throughout training (pressure stability test — flag if field_mse goes to Infinity or NaN at any epoch)
- Whether it beats baseline 0.002383
- Best epoch achieved

## Baseline

Current best metrics (TandemFoil Paper):
- **val_primary/field_mse: 0.002383** (best at epoch 443)
- val/surface_mse: 0.001517
- val/surface_mse_p: 4.81e-05
- val/volume_mse: 0.002397
- Baseline PR: #3025 (haku — 3L/192d, Lion lr=1.25e-4, T_max=10, gc=0.5, EMA=0.999)
- Baseline W&B run: d1xh0o1p (haku/tfp-lion-champion-config)
- Reproduce command:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nami:

## Results

### Experiment: 3L/224d TandemFoil Paper width sweep

**Outcome: FAILED — model diverged catastrophically at epoch 87.**

| Metric | Baseline (3L/192d) | 3L/224d (best ep 46) | Beat baseline? |
|--------|-------------------|---------------------|----------------|
| val_primary/field_mse | 0.002383 | 2.20e9* | No |
| val/surface_mse_Ux | — | 0.00248 | — |
| val/surface_mse_Uy | — | 0.00665 | — |
| val/surface_mse_p | — | 9.29e8* | — |
| val/volume_mse_Ux | — | 0.01670 | — |
| val/volume_mse_Uy | — | 0.01277 | — |
| val/volume_mse_p | — | 6.68e9* | — |

*\*Pressure metrics are inflated — see bug note below.*

- **Best epoch:** 46 (of 102 before killed; diverged at epoch 87)
- **W&B run:** `g3jp3nfi` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/g3jp3nfi))
- **Pressure stability:** Pressure **never stabilized** to a comparable range. Velocity channels converged well (surface_mse_Ux=0.0025, Uy=0.0067), but pressure dominated field_mse.

### What happened

1. **Epochs 1–86: Healthy training.** Loss decreased steadily from 1.01 to 0.06. Velocity MSE dropped nicely. Grad norms settled from ~25 to ~11. All grad clip events (~360/360 batches clipped throughout).

2. **Epoch 87: Gradient spike.** Grad norm jumped from 11.1 to 48.0 (4x). Loss was still 0.06.

3. **Epoch 88: Catastrophic divergence.** Grad norm exploded to 4.03e8. Loss jumped from 0.06 to 3.91. Model weights were corrupted beyond recovery.

4. **Epochs 89–102: Dead model.** Grad norm = inf, loss stuck at ~3.06. Killed the run.

### Analysis

**224d is unstable with the 192d champion hyperparameters.** The 192d baseline trained stably for 443+ epochs; the 224d model diverged at epoch 87. The wider model has ~37% more parameters (224²×3 vs 192²×3), and the same grad_clip=0.5 + Lion lr=1.25e-4 configuration cannot stabilize it through the cosine cycles.

The divergence point (epoch 87, mid-cycle 9 of T_max=10) was NOT at a cosine restart — it was a spontaneous gradient explosion, suggesting the model was in a fragile equilibrium that one pathological batch was enough to break.

This provides evidence that **192d is at or near the stability sweet spot** for 3L models with the champion hyperparameters. Width is not a reliable lever to push performance further without also adjusting the training dynamics (lower lr, stronger clipping, warmup, etc.).

### Train command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 \
  --model-hidden-dim 224 --model-heads 4 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent nami --wandb-name "nami/tfp-3l-224d-v3" \
  --wandb-group tfp-width-sweep
```

### Bug fixes applied to train.py

Two bugs were found and fixed during this experiment:

1. **`primary_metric_key` variable shadowing (line 1641):** A local variable `primary_metric_key` was assigned on lines 1641-1643, shadowing the function `primary_metric_key()` defined at line 1423. Python's scoping rules caused an `UnboundLocalError` at line 1634 where the function was called. Fixed by renaming the local variable to `primary_metric_str`.

2. **`sinh` overflow in `TargetTransform.invert()` (line 230):** The `sinh` inverse for asinh-encoded pressure had no clamp, causing float32 overflow to Infinity. The normalization stats (mean=-23.5, std=462) are computed on raw physical pressure, but the normalized training targets use asinh compression, creating a mismatch where any prediction with un-normalized value outside [-20, 20] overflows `sinh`. Fixed by clamping to [-20, 20] before `sinh`, matching the `TandemTargetTransform._phys_denorm` clamp range. Note: the clamped pressure metrics are not directly comparable to baseline numbers from runs without this fix, but at convergence (when model predictions are accurate) the clamp has zero effect.

### Suggested follow-ups

- **Lower learning rate for 224d:** Try lr=8e-5 or 6e-5 with the wider model to prevent gradient explosion.
- **Stronger grad clipping:** Try grad_clip=0.3 or 0.25 for 224d to provide more stability margin.
- **The sinh overflow bug** should be fixed in the main codebase — it affects all tandemfoil_paper evaluation when asinh_pressure is enabled. The proper long-term fix is to recompute the split stats JSON with asinh-compressed pressure values so the normalization stats match the transform domain.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3157 because ADVISOR: 3L/224d catastrophically diverges — field_mse ~2.2e9, gradient explosion at ep87 (grad norm →4e8). Combined with 4L/192d pressure overflow and 3L/256d (robin, in-progress), this confirms 192d as the stability sweet spot. Width beyond 192d amplifies gradient instability in TFP's narrow stability window. The sinh overflow bug fix you identified is valuable — please submit it as a separate bugfix PR. Reassigning you to AirfRANS where we have a meaningful volume gap to close..
